tools/shared/src/main/scala/org/scalajs/core/tools/linker/backend/emitter/FunctionEmitter.scala

```diff
@@ -278,23 +362,35 @@ private[emitter] class FunctionEmitter(jsGen: JSGen) {
      */
     def desugarToFunctionWithExplicitThis(
         params: List[ParamDef], body: Tree, isStat: Boolean, env0: Env)(
-        implicit pos: Position): js.Function = {
+        implicit pos: Position): WithGlobals[js.Function] = {
 
-      /* TODO The identifier `$thiz` cannot be produced by 0.6.x compilers due
-       * to their name mangling, which guarantees that it is unique. We should
-       * find a better way to do this in the future, though.
-       */
-      val thisIdent = js.Ident("$thiz", Some("this"))
-      val env = env0.withThisIdent(Some(thisIdent))
-      val js.Function(jsParams, jsBody) =
-        desugarToFunction(params, body, isStat, env)
-      js.Function(js.ParamDef(thisIdent, rest = false) :: jsParams, jsBody)
+      performOptimisticThenPessimisticRuns {
+        /* TODO The identifier `$thiz` cannot be produced by 0.6.x compilers due
+         * to their name mangling, which guarantees that it is unique. We should
+         * find a better way to do this in the future, though.
+         */
```

Comment:
Filed as #2972. IIUC, this is required for correctness. Otherwise someone could try to refer to the global name `$thiz` and boom!